x-pack/plugins/lens/public/xy_visualization_plugin/xy_config_panel.tsx

```diff
@@ -0,0 +1,250 @@
+  },
+];
+
+export function XYConfigPanel(props: VisualizationProps<State>) {
+  const { state, datasource, setState } = props;
+
+  return (
+    <EuiForm className="lnsConfigPanel">
+      <EuiFormRow label="Chart type">
+        <EuiButtonGroup
+          legend="Chart Type"
+          name="chartType"
+          className="eui-displayInlineBlock"
+          data-test-subj="lnsXY_seriesType"
+          options={chartTypeIcons}
+          idSelected={state.seriesType}
+          onChange={seriesType => {
+            const isHorizontal = seriesType === 'horizontal_bar';
+            setState({
+              ...state,
+              seriesType: seriesType as SeriesType,
+              x: {
+                ...state.x,
+                position: isHorizontal ? Position.Left : Position.Bottom,
+              },
+              y: {
+                ...state.y,
+                position: isHorizontal ? Position.Bottom : Position.Left,
+              },
+            });
+          }}
+          isIconOnly
+        />
+      </EuiFormRow>
+
+      <EuiFormRow label="Title">
+        <EuiFieldText
+          placeholder="Title"
+          data-test-subj="lnsXY_title"
+          value={state.title}
+          onChange={e => setState({ ...state, title: e.target.value })}
+          aria-label="Title"
+        />
+      </EuiFormRow>
+
+      <EuiFormRow>
+        <EuiSwitch
+          label="Show legend"
+          checked={state.legend.isVisible}
+          data-test-subj="lnsXY_legendIsVisible"
+          onChange={() =>
+            setState({ ...state, legend: { ...state.legend, isVisible: !state.legend.isVisible } })
+          }
+        />
+      </EuiFormRow>
+
+      {state.legend.isVisible && (
+        <EuiFormRow label="Legend position">
+          <EuiButtonGroup
+            legend="Legend position"
+            data-test-subj="lnsXY_legendPosition"
+            name="legendPosition"
+            options={positionIcons}
+            idSelected={state.legend.position}
+            onChange={position =>
+              setState({ ...state, legend: { ...state.legend, position: position as Position } })
+            }
+            isIconOnly
+          />
+        </EuiFormRow>
+      )}
+
+      <EuiFormRow label="X Axis">
+        <>
+          <EuiFormRow label="Title">
+            <EuiFieldText
+              placeholder="Title"
+              data-test-subj="lnsXY_xTitle"
+              value={state.x.title}
+              onChange={e => setState({ ...state, x: { ...state.x, title: e.target.value } })}
+              aria-label="Title"
+            />
+          </EuiFormRow>
+
+          <EuiFormRow label="Value">
+            <div
+              data-test-subj="lnsXY_xDimensionPanel"
+              ref={el =>
+                el &&
+                datasource.renderDimensionPanel(el, {
+                  columnId: state.x.accessor,
+                  filterOperations: () => true,
+                })
+              }
+            />
```

Comment:
The callback ref here is exactly what the `NativeRenderer` is doing- did you try that, and did it work?

Reply:
Nope. I forgot we had that. I'll drop it in. The one annoying thing about that is that it requires me to explicitly define some of the type signature in the call.